File: Exemplos/App_IntentDigitalHUB_Flutter/Elgin_IntentDigitalHUB_Flutter-font/lib/IntentDigitalHubService/intent_digital_hub_command.dart
part of 'intent_digital_hub_command_starter.dart';

abstract class IntentDigitalHubCommand {
  final String _functionName;
  final IntentDigitalHubModule _correspondingHubModule;

  IntentDigitalHubCommand(this._functionName, this._correspondingHubModule);

  //JSON do comando formatado
  String get _commandJSON {
    return '[{'
            '"funcao"'
            ':'
            '"' +
        _functionName +
        '"'
            ','
            '"parametros"'
            ':'
            '{' +
        functionParameters() +
        '}'
            '}]';
  }

  //Função que deve ser implementada pelas subsclasses definindo a formatação do JSON dos paramêtros
  @protected
  String functionParameters();
}

File: Exemplos/App_IntentDigitalHUB_XamarinForms/Elgin_IntentDigitalHUB_XamarinForms-font/E1_Digital_Hub_Xamarin_Forms_M8/IntentServices/IntentDigitalHubCommand.cs
using Newtonsoft.Json;

namespace Xamarin_Forms_Intent_Digital_Hub.IntentServices
{
    public abstract class IntentDigitalHubCommand
    {
        protected string functionName;
        public IntentDigitalHubModule correspondingIntentModule;

        [JsonProperty(PropertyName = "funcao")]
        private string funcao;

        [JsonProperty(PropertyName = "mensagem")]
        private string mensagem;

        //O código de sucesso de um comando no DigitalHub é 0
        readonly static public int RESULT_OK = 0;

        protected IntentDigitalHubCommand(string functionName, IntentDigitalHubModule correspondingIntentModule)
        {
            this.functionName = functionName;
            this.correspondingIntentModule = correspondingIntentModule;
        }

        //Formata o JSON de acordo com os parâmetros definidos por cada subclasse, o modificador de acesso protected impede a exposição do método que sera usado somente em DigitalHubUtils para o start da intent
        public string GetCommandJSON()
        {
            return "[{" +
                    "\"funcao\"" + ":" + "\"" + functionName + "\"" + "," +
                    "\"parametros\"" + ":" + "{" + FunctionParameters() + "}" +
                    "}]";
        }

        //Função que deve ser implementada por cada subclasse definindo a formatação dos parâmetros da função específica
        protected abstract string FunctionParameters();

        //Getters do Serializable
        public string GetRetornoFuncao()
        {
            return funcao;
        }

        public string GetRetornoMensagem()
        {
            return mensagem;
        }
    }
}
